classes/class-gf-securesubmit.php

```diff
@@ -1,950 +1,1597 @@
-            return $response->error->message;
-        }
-
-        return false;
-    }
-
-    public function isFieldOnValidPage($field, $parent)
-    {
-        $form = $this->get_current_form();
-
-        $mapped_field_id   = $this->get_setting($field['name']);
-        $mapped_field      = GFFormsModel::get_field($form, $mapped_field_id);
-        $mapped_field_page = rgar($mapped_field, 'pageNumber');
-
-        $cc_field = $this->get_credit_card_field($form);
-        $cc_page  = rgar($cc_field, 'pageNumber');
-
-        if ($mapped_field_page > $cc_page) {
-            $this->set_field_error($field, __('The selected field needs to be on the same page as the Credit Card field or a previous page.', 'gravityforms-securesubmit'));
-        }
-    }
-
-    public function replaceMergeTags($text, $form, $entry, $url_encode, $esc_html, $nl2br, $format)
-    {
-        $mergeTags = array(
-            'transactionId'     => '{securesubmit_transaction_id}',
-            'authorizationCode' => '{securesubmit_authorization_code}'
-        );
-
-        $gFormsKey = array(
-            'transactionId' => 'transaction_id',
-        );
-
-        foreach ($mergeTags as $key => $mergeTag) {
-            // added for GF 1.9.x
-            if (strpos($text, $mergeTag) === false || empty($entry) || empty($form)) {
-                return $text;
-            }
-
-            $value = '';
-            if (class_exists('GFSecureSubmit') && isset(GFSecureSubmit::get_instance()->transaction_response)) {
-                $value = GFSecureSubmit::get_instance()->transaction_response->$key;
-            }
-
-            if (isset($gFormsKey[$key]) && empty($value)) {
-                $value = rgar($entry, $gFormsKey[$key]);
-            }
-
-            $text = str_replace($mergeTag, $value, $text);
-        }
-        return $text;
-    }
-
-    public function addClientSideMergeTags($form)
-    {
-        include plugin_dir_path(__FILE__) . '../templates/client-side-merge-tags.php';
-        return $form;
-    } 
-    /**Attempts to get real ip even if there is a proxy chain
-     * @return string
-     */
-    private function getRemoteIP() {
-        $remoteIP = $_SERVER['REMOTE_ADDR'];
-        if (array_key_exists('HTTP_X_FORWARDED_FOR', $_SERVER) && $_SERVER['HTTP_X_FORWARDED_FOR'] != '') {
-            $remoteIPArray = array_values(
-                array_filter(
-                    explode(
-                        ',',
-                        $_SERVER['HTTP_X_FORWARDED_FOR']
-                    )
-                )
-            );
-            $remoteIP = end($remoteIPArray);
-        }
-        return $remoteIP;
-    }
-}
+<?php
+GFForms::include_payment_addon_framework();
+include_once 'class-gf-field-hpsach.php';
+
+/**
+ * Handles Heartlands Payments with Gravity Forms
+ * Class GFSecureSubmit
+ */
+class GFSecureSubmit
+    extends GFPaymentAddOn {
+    /**
+     * @var bool
+     */
+    private $isCC = null;
```

Comment:
Private bool that starts as a null? This should default to false if it's going to be an instance variable.